src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/StringSearchValuesHelper.cs

```diff
@@ -0,0 +1,294 @@
+                    return differentBits == 0;
+                }
+                else
+                {
+                    nuint offset = byteLength - sizeof(uint);
+                    uint differentBits = (Unsafe.ReadUnaligned<uint>(ref first) & ~0x200020u) - Unsafe.ReadUnaligned<uint>(ref second);
+                    differentBits |= (Unsafe.ReadUnaligned<uint>(ref Unsafe.Add(ref first, offset)) & ~0x200020u) - Unsafe.ReadUnaligned<uint>(ref Unsafe.Add(ref second, offset));
+                    return differentBits == 0;
+                }
+            }
+        }
+
+        public readonly struct CaseInsensitiveAscii : ICaseSensitivity
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static char TransformInput(char input) => TextInfo.ToUpperAsciiInvariant(input);
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector128<byte> TransformInput(Vector128<byte> input)
+            {
+                Vector128<byte> subtraction = Vector128.Create((byte)(128 + 'a'));
+                Vector128<byte> comparison = Vector128.Create((byte)(128 + 26));
+                Vector128<byte> caseConversion = Vector128.Create((byte)0x20);
+
+                Vector128<byte> matches = Vector128.LessThan((input - subtraction).AsSByte(), comparison.AsSByte()).AsByte();
+                return input ^ (matches & caseConversion);
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector256<byte> TransformInput(Vector256<byte> input)
+            {
+                Vector256<byte> subtraction = Vector256.Create((byte)(128 + 'a'));
+                Vector256<byte> comparison = Vector256.Create((byte)(128 + 26));
+                Vector256<byte> caseConversion = Vector256.Create((byte)0x20);
+
+                Vector256<byte> matches = Vector256.LessThan((input - subtraction).AsSByte(), comparison.AsSByte()).AsByte();
+                return input ^ (matches & caseConversion);
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector512<byte> TransformInput(Vector512<byte> input)
+            {
+                Vector512<byte> subtraction = Vector512.Create((byte)(128 + 'a'));
+                Vector512<byte> comparison = Vector512.Create((byte)(128 + 26));
+                Vector512<byte> caseConversion = Vector512.Create((byte)0x20);
+
+                Vector512<byte> matches = Vector512.LessThan((input - subtraction).AsSByte(), comparison.AsSByte()).AsByte();
+                return input ^ (matches & caseConversion);
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static bool Equals(ref char matchStart, string candidate)
```

Comment:
We don't already have helpers for doing this elsewhere in corelib?  This all feels like a lot of code duplication... is it in order to eek out a few additional cycles?

There's _a lot_ of code involved in all of this.  It'd be nice to pull back a bit on all the duplication.

Reply:
I removed the `IValueLength` specializations here as they didn't end up impacting perf significantly enough. With that, the amount of logic here is greatly reduced.

The main things that remain are the vector Ascii ToUpper conversions that assume the input is already Ascii, I don't think we have exactly that sort of helper elsewhere.